src/kernels/MIOpenSparseSoftmaxCrossEntropyWithLogits.cpp

```diff
@@ -0,0 +1,280 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2024 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+#ifndef T_TYPE
+#define T_TYPE int
+#endif
+
+#ifndef DTYPE_LOWEST
+#define DTYPE_LOWEST 0
+#endif
+
+#include "float_types.h"
+#include "tensor_view.hpp"
+
+template <typename T, typename Ta>
+__device__ void sparseSoftmaxCrossEntropyWithLogitsForward(const T* input,
+                                                           const Ta* target,
+                                                           T* output,
+                                                           T* backprop,
+                                                           uint64_t num_class,
+                                                           tensor_view_t<2> input_tv,
+                                                           tensor_view_t<1> target_tv,
+                                                           tensor_view_t<1> output_tv,
+                                                           tensor_view_t<2> backprop_tv)
+{
+    uint64_t gid = blockIdx.x;
+    uint64_t lid = threadIdx.x;
+
+    __shared__ FLOAT_ACCUM lmax[LOCAL_SIZE], lsum[LOCAL_SIZE];
```

Comment:
Those variable may not be needed to be `shared` array if you use template code at `block_reduce.hpp`.

Reply:
resolved